### PR TITLE
update Schnorr signature logic to match BIP340 spec/reference impleme…

### PR DESCRIPTION
**Pull request by bflood366**
…ntation

### CODE REVIEW

File: integrationtests/src/test/kotlin/xyz/funkybit/integrationtests/bitcoin/UtxoSelectionTest.kt [hunk 1]
```diff
@@ -38,7 +39,7 @@ class UtxoSelectionTest {
 
     @Test
     fun testUxtoSelection() {
-        Assumptions.assumeTrue((System.getenv("INTEGRATION_RUN") ?: "0") != "1")
+        Assumptions.assumeFalse(isTestEnvRun())
```
Comment (bflood366) [inline integrationtests/src/test/kotlin/xyz/funkybit/integrationtests/bitcoin/UtxoSelectionTest.kt:42]: this is to fix an issue with my last push to master - these tests should not run yet test env.

File: backend/src/main/kotlin/xyz/funkybit/core/utils/schnorr/Point.kt [hunk 2]
```diff
@@ -19,14 +19,7 @@ class Point(x: BigInteger?, y: BigInteger?) {
...
         fun taggedHash(tag: String, msg: ByteArray): ByteArray {
-            val tagHash: ByteArray = sha256(tag.toByteArray())
-            val len = (tagHash.size * 2) + msg.size
-            val buf = ByteArray(len)
-            System.arraycopy(tagHash, 0, buf, 0, tagHash.size)
-            System.arraycopy(tagHash, 0, buf, tagHash.size, tagHash.size)
-            System.arraycopy(msg, 0, buf, tagHash.size * 2, msg.size)
-
-            return sha256(buf)
+            return sha256(sha256(tag.toByteArray()) + sha256(tag.toByteArray()) + msg)
```
Comment (bholzman) [inline backend/src/main/kotlin/xyz/funkybit/core/utils/schnorr/Point.kt:22]: might as well only calculate `sha256(tag.toByteArray())` once

File: backend/src/main/kotlin/xyz/funkybit/core/utils/schnorr/Schnorr.kt [hunk 3]
```diff
@@ -1,86 +1,149 @@
...
+        val p = Point.G.mul(d0)!!
+
+        // 4. Let d = d' if has_even_y(P), otherwise let d = n - d'
+        val d = if (p.hasEvenY()) d0 else Point.n.subtract(d0)
+
+        // 5. Let t be the byte-wise xor of bytes(d) and hashBIP0340/aux(a)[11].
+        val t = xorBytes(d.toByteArrayNoSign(), Point.taggedHash("BIP0340/aux", auxRand))
+
+        // 6. Let rand = hashBIP0340/nonce(t || bytes(P) || m)[12].
+        val rand = Point.taggedHash("BIP0340/nonce", t + Point.bytesFromPoint(p) + m)
```
Comment (bholzman) [inline backend/src/main/kotlin/xyz/funkybit/core/utils/schnorr/Schnorr.kt:53]: ah, so these two steps were the main problem

File: backend/src/main/kotlin/xyz/funkybit/core/utils/schnorr/Schnorr.kt [hunk 4]
```diff
@@ -1,86 +1,149 @@
...
-        val k: BigInteger? = if (!r.hasSquareY()) {
-            Point.n.subtract(k0)
-        } else {
-            k0
+
+        // 10. Let k = k' if has_even_y(R), otherwise let k = n - k' .
+        val k = if (r.hasEvenY()) k0 else Point.n.subtract(k0)
+
+        // 11. Let e = int(hashBIP0340/challenge(bytes(R) || bytes(P) || m)) mod n
+        val e = BigInteger(1, Point.taggedHash("BIP0340/challenge", Point.bytesFromPoint(r) + Point.bytesFromPoint(p) + m)).mod(Point.n)
```
Comment (bholzman) [inline backend/src/main/kotlin/xyz/funkybit/core/utils/schnorr/Schnorr.kt:70]: ah, and this one

**Review by bholzman — APPROVED** [review]
Looks great, nice work!

File: backend/src/main/kotlin/xyz/funkybit/core/utils/schnorr/Point.kt (hunk 2 above)
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/core/utils/schnorr/Point.kt]: yep changed

File: backend/src/main/kotlin/xyz/funkybit/core/utils/schnorr/Schnorr.kt (hunk 4 above)
Comment (bflood366) [reply backend/src/main/kotlin/xyz/funkybit/core/utils/schnorr/Schnorr.kt:70]: yeah - it was pretty close - there was also a few minor changes needed in Point.kt - luckily found a python reference implementation  in that bitcoin repo so was able to print values after each step and compare till I got it working.

**Pull request merged**